integrations/google_ai/src/haystack_integrations/components/generators/google_ai/chat/gemini.py

```diff
@@ -336,11 +344,26 @@ def _get_stream_response(
         :param streaming_callback: The handler for the streaming response.
         :returns: The extracted response with the content of all streaming chunks.
         """
-        responses = []
+        replies: List[ChatMessage] = []
+        content: Union[str, Dict[str, Any]] = ""
         for chunk in stream:
-            content = chunk.text if len(chunk.parts) > 0 and "text" in chunk.parts[0] else ""
-            streaming_callback(StreamingChunk(content=content, meta=chunk.to_dict()))
-            responses.append(content)
+            metadata = chunk.to_dict()  # we store whole chunk as metadata in streaming calls
+            for candidate in chunk.candidates:
+                for part in candidate.content.parts:
+                    if part.text != "":
+                        content = part.text
+                        replies.append(ChatMessage(content=content, role=ChatRole.ASSISTANT, meta=metadata, name=None))
+                    elif part.function_call is not None:
+                        metadata["function_call"] = part.function_call
+                        content = dict(part.function_call.args.items())
+                        replies.append(
+                            ChatMessage(
+                                content=content,
+                                role=ChatRole.ASSISTANT,
+                                name=part.function_call.name,
+                                meta=metadata,
+                            )
+                        )
 
-        combined_response = "".join(responses).lstrip()
-        return [ChatMessage.from_assistant(content=combined_response)]
+                streaming_callback(StreamingChunk(content=content, meta=metadata))
```

Comment:
Is the indentation correct here? We're calling `streaming_callback` once per `candidate`. Shouldn't it be called once per `part`? What if there are multiple `parts` per `candidate`?

Reply:
You have a valid point. I looked into this, but it's difficult to find any chunk that contains multiple Parts, so for now, it produces the same result. In other generators, the entire chunk is passed to the callback function, as shown [here](https://github.com/deepset-ai/haystack-core-integrations/blob/9fa18533fa52cd56ce1676d13def4113ade9df4f/integrations/amazon_bedrock/src/haystack_integrations/components/generators/amazon_bedrock/adapters.py#L57). Because of this, I'm not entirely certain, we can do it your way.
@anakin87 would you have an input here?

Reply:
Sorry. @vblagoje knows better.

Reply:
@silvanocerza I updated based on your suggestion for now which makes more sense, also this PR is blocking my other PR. So I'll just go ahead